File: implementations/python/map1/_projection.py
"""MAP v1.1 projection — FULL and BIND modes.

FULL (§2.2): the identity function on the descriptor MAP.
BIND (§2.3): select specific fields by RFC 6901 JSON Pointer paths,
             producing a minimal enclosing MAP structure.

BIND is where most of the complexity lives.  The spec has five pointer-set
rules (a–e) plus four structural rules (1–4) that interact in non-obvious
ways.  The comments below reference specific spec rules so implementers
can trace each branch back to normative text.
"""

from __future__ import annotations

from typing import Any, Dict, List, Tuple

from ._errors import ERR_SCHEMA, MapError


# ── RFC 6901 JSON Pointer parsing ─────────────────────────────
# RFC 6901 is simple but has one sharp edge: tilde escaping.
# "~0" → literal "~" and "~1" → literal "/".  The order matters —
# if you decode "~1" before "~0", the string "~01" decodes wrong.
# We handle this character-by-character to avoid that trap.

def _parse_pointer(ptr: str) -> List[str]:
    """Parse an RFC 6901 pointer into reference tokens.

    "" (empty string) → [] (whole-document pointer, rule 2.3.e).
    Otherwise must start with "/".
    """
    if ptr == "":
        return []
    if not ptr.startswith("/"):
        raise MapError(ERR_SCHEMA, "pointer must start with '/'")

    tokens: List[str] = []
    for raw in ptr.split("/")[1:]:
        decoded = ""
        i = 0
        while i < len(raw):
            if raw[i] != "~":
                decoded += raw[i]
                i += 1
                continue
            # Must have a character after ~
            if i + 1 >= len(raw):
                raise MapError(ERR_SCHEMA, "dangling ~ in pointer")
            nxt = raw[i + 1]
            if nxt == "0":
                decoded += "~"
            elif nxt == "1":
                decoded += "/"
            else:
                raise MapError(ERR_SCHEMA, "bad ~{} escape in pointer".format(nxt))
            i += 2
        tokens.append(decoded)
    return tokens


# ── FULL projection (§2.2) ────────────────────────────────────

def full_project(descriptor: Any) -> Any:
    """FULL projection: identity on the descriptor."""
    return descriptor


# ── BIND projection (§2.3) ────────────────────────────────────

def bind_project(descriptor: Any, pointers: List[str]) -> Any:
    """BIND projection: select fields by JSON Pointer paths.

    Implements all normative rules from §2.3:
      (a) Parse every pointer per RFC 6901
      (b) Reject duplicate pointers
      (c) Unmatched pointer handling (fail-closed unless zero match)
      (d) Subsumption of overlapping pointers
      (e) Empty pointer "" = FULL-equivalent
      (1) Omit siblings at each MAP level
      (2) Minimal enclosing structure
      (3) No match → empty MAP (not an error)
      (4) LIST traversal is forbidden (ERR_SCHEMA)
    """
    # Root must be a MAP.
    if not isinstance(descriptor, dict):
        raise MapError(ERR_SCHEMA, "BIND root must be a MAP")

    # Rule (b): no duplicate pointer strings.
    if len(set(pointers)) != len(pointers):
        raise MapError(ERR_SCHEMA, "duplicate pointers")

    # Rule (a): parse all pointers up front so parse failures are caught
    # before we start traversing the descriptor.
    parsed: List[Tuple[str, List[str]]] = []
    for ptr in pointers:
        tokens = _parse_pointer(ptr)
        parsed.append((ptr, tokens))

    # Walk each pointer against the descriptor to determine match status.
    matched_paths: List[List[str]] = []
    any_match = False
    any_unmatched = False

    for ptr, tokens in parsed:
        # Rule (e): empty pointer always matches the MAP root.
        if ptr == "":
            any_match = True
            continue

        cur: Any = descriptor
        ok = True
        for tok in tokens:
            # Rule (4): LIST traversal is forbidden.
            if isinstance(cur, list):
                raise MapError(ERR_SCHEMA, "BIND cannot traverse LIST")
            if not isinstance(cur, dict) or tok not in cur:
                ok = False
                break
            cur = cur[tok]

        if ok:
            any_match = True
            matched_paths.append(tokens)
        else:
            any_unmatched = True

    # Rule (c): unmatched pointer handling.
    if not any_match:
        return {}  # Rule (3): all pointers unmatched → empty MAP
    if any_unmatched:
        # At least one matched but another didn't → fail-closed.
        raise MapError(ERR_SCHEMA, "unmatched pointer in set")

    # Rule (e): if any pointer is "", result is the full descriptor.
    if any(ptr == "" for ptr, _ in parsed):
        return descriptor

    # Rule (d): discard subsumed pointers (P1 is prefix of P2 → P2 is redundant).
    def is_subsumed(toks: List[str]) -> bool:
        for other in matched_paths:
            if len(other) < len(toks) and toks[:len(other)] == other:
                return True
        return False

    effective = [t for t in matched_paths if not is_subsumed(t)]

    # Build the projected tree — rule (1) omit-siblings, rule (2) minimal structure.
    projected: Dict[str, Any] = {}
    for toks in effective:
        cur = descriptor
        path_keys: List[Tuple[dict, str]] = []
        for tok in toks:
            if isinstance(cur, list):
                raise MapError(ERR_SCHEMA, "BIND cannot traverse LIST")
            if not isinstance(cur, dict):
                raise MapError(ERR_SCHEMA, "cannot traverse non-MAP")
            path_keys.append((cur, tok))
            cur = cur[tok]

        # Walk the projected tree, creating nested MAPs as needed.
        target = projected
        for i, (_node, tok) in enumerate(path_keys):
            if i == len(path_keys) - 1:
                target[tok] = cur  # leaf value
            else:
                nxt = target.get(tok)
                if nxt is None:
                    nxt = {}
                    target[tok] = nxt
                if not isinstance(nxt, dict):
                    raise MapError(ERR_SCHEMA, "BIND path conflict")
                target = nxt

    return projected

File: implementations/go/projection.go
package map1

import "strings"

// FullProject returns the descriptor unchanged (§2.2).
func FullProject(descriptor Value) Value {
	return descriptor
}

// BindProject selects fields from descriptor by JSON Pointer paths (§2.3).
//
// Implements all normative rules:
//
//	(a) Parse every pointer per RFC 6901
//	(b) Reject duplicate pointers
//	(c) Unmatched pointer handling (fail-closed unless zero match)
//	(d) Subsumption of overlapping pointers
//	(e) Empty pointer "" = FULL-equivalent
//	(1) Omit siblings at each MAP level
//	(2) Minimal enclosing structure
//	(3) No match → empty MAP
//	(4) LIST traversal forbidden (ERR_SCHEMA)
func BindProject(descriptor Value, pointers []string) (Value, error) {
	// Root must be a MAP.
	root, ok := descriptor.(*Map)
	if !ok {
		return nil, newErr(ErrSchema, "BIND root must be a MAP")
	}

	// Rule (b): no duplicate pointer strings.
	seen := make(map[string]bool, len(pointers))
	for _, p := range pointers {
		if seen[p] {
			return nil, newErr(ErrSchema, "duplicate pointers")
		}
		seen[p] = true
	}

	// Rule (a): parse all pointers up front.
	type parsedPtr struct {
		raw    string
		tokens []string
	}
	parsed := make([]parsedPtr, len(pointers))
	for i, ptr := range pointers {
		tokens, err := parsePointer(ptr)
		if err != nil {
			return nil, err
		}
		parsed[i] = parsedPtr{raw: ptr, tokens: tokens}
	}

	// Walk each pointer to determine match status.
	type matchedPath struct {
		tokens []string
	}
	var matched []matchedPath
	anyMatch := false
	anyUnmatched := false

	for _, pp := range parsed {
		// Rule (e): empty pointer always matches.
		if pp.raw == "" {
			anyMatch = true
			continue
		}
		cur := Value(root)
		ok := true
		for _, tok := range pp.tokens {
			// Rule (4): LIST traversal forbidden.
			if _, isList := cur.(List); isList {
				return nil, newErr(ErrSchema, "BIND cannot traverse LIST")
			}
			m, isMap := cur.(*Map)
			if !isMap {
				ok = false
				break
			}
			found := false
			for j, k := range m.Keys {
				if k == tok {
					cur = m.Values[j]
					found = true
					break
				}
			}
			if !found {
				ok = false
				break
			}
		}
		if ok {
			anyMatch = true
			matched = append(matched, matchedPath{tokens: pp.tokens})
		} else {
			anyUnmatched = true
		}
	}

	// Rule (c): unmatched pointer handling.
	if !anyMatch {
		return EmptyMap(), nil // Rule (3)
	}
	if anyUnmatched {
		return nil, newErr(ErrSchema, "unmatched pointer in set")
	}

	// Rule (e): if any pointer is "", result is full descriptor.
	for _, pp := range parsed {
		if pp.raw == "" {
			return root, nil
		}
	}

	// Rule (d): discard subsumed pointers.
	effective := make([][]string, 0, len(matched))
	for _, mp := range matched {
		subsumed := false
		for _, other := range matched {
			if len(other.tokens) < len(mp.tokens) {
				if tokensPrefix(other.tokens, mp.tokens) {
					subsumed = true
					break
				}
			}
		}
		if !subsumed {
			effective = append(effective, mp.tokens)
		}
	}

	// Build projected tree — rule (1) omit-siblings, rule (2) minimal structure.
	projected := &Map{}
	for _, toks := range effective {
		cur := Value(root)
		// Resolve the leaf value.
		for _, tok := range toks {
			m := cur.(*Map) // safe — we already validated the path
			for j, k := range m.Keys {
				if k == tok {
					cur = m.Values[j]
					break
				}
			}
		}
		leaf := cur

		// Walk the projected tree, creating nested Maps as needed.
		target := projected
		for i, tok := range toks {
			if i == len(toks)-1 {
				// Leaf assignment.
				mapSet(target, tok, leaf)
			} else {
				existing := mapGet(target, tok)
				if existing == nil {
					child := &Map{}
					mapSet(target, tok, child)
					target = child
				} else {
					child, ok := existing.(*Map)
					if !ok {
						return nil, newErr(ErrSchema, "BIND path conflict")
					}
					target = child
				}
			}
		}
	}

	return projected, nil
}

// parsePointer parses an RFC 6901 JSON Pointer into reference tokens.
// "" → [] (whole-document pointer, rule 2.3.e).
func parsePointer(ptr string) ([]string, error) {
	if ptr == "" {
		return nil, nil
	}
	if !strings.HasPrefix(ptr, "/") {
		return nil, newErr(ErrSchema, "pointer must start with '/'")
	}
	parts := strings.Split(ptr[1:], "/")
	tokens := make([]string, len(parts))
	for i, raw := range parts {
		// RFC 6901 tilde decoding: ~0 → "~", ~1 → "/".
		// Process character-by-character to handle ~01 correctly.
		var b strings.Builder
		j := 0
		for j < len(raw) {
			if raw[j] != '~' {
				b.WriteByte(raw[j])
				j++
				continue
			}
			if j+1 >= len(raw) {
				return nil, newErr(ErrSchema, "dangling ~ in pointer")
			}
			switch raw[j+1] {
			case '0':
				b.WriteByte('~')
			case '1':
				b.WriteByte('/')
			default:
				return nil, newErr(ErrSchema, "bad tilde escape in pointer")
			}
			j += 2
		}
		tokens[i] = b.String()
	}
	return tokens, nil
}

// tokensPrefix returns true if a is a strict prefix of b.
func tokensPrefix(a, b []string) bool {
	if len(a) >= len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// Map helpers — these operate on our Map type (not Go's map).
func mapGet(m *Map, key string) Value {
	for i, k := range m.Keys {
		if k == key {
			return m.Values[i]
		}
	}
	return nil
}

func mapSet(m *Map, key string, val Value) {
	for i, k := range m.Keys {
		if k == key {
			m.Values[i] = val
			return
		}
	}
	m.Keys = append(m.Keys, key)
	m.Values = append(m.Values, val)
}
